File: tests/cpp-tests/Classes/TerrainTest/TerrainTest.cpp
/****************************************************************************
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.
 
 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "TerrainTest.h"

USING_NS_CC;

TerrainTests::TerrainTests()
{
    ADD_TEST_CASE(TerrainSimple);
    ADD_TEST_CASE(TerrainWalkThru);
    ADD_TEST_CASE(TerrainWithLightMap);
}

Vec3 camera_offset(0, 45, 60);
#define PLAYER_HEIGHT 0

TerrainSimple::TerrainSimple()
{
    Size visibleSize = Director::getInstance()->getVisibleSize();

    //use custom camera
    _camera = Camera::createPerspective(60,visibleSize.width/visibleSize.height,0.1f,800);
    _camera->setCameraFlag(CameraFlag::USER1);
    _camera->setPosition3D(Vec3(-1,1.6f,4));
    addChild(_camera);

    Terrain::DetailMap r("TerrainTest/dirt.jpg"),g("TerrainTest/Grass2.jpg"),b("TerrainTest/road.jpg"),a("TerrainTest/GreenSkin.jpg");

    Terrain::TerrainData data("TerrainTest/heightmap16.jpg","TerrainTest/alphamap.png",r,g,b,a);

    _terrain = Terrain::create(data,Terrain::CrackFixedType::SKIRT);
    _terrain->setLODDistance(3.2f,6.4f,9.6f);
    _terrain->setMaxDetailMapAmount(4);
    addChild(_terrain);
    _terrain->setCameraMask(2);
    _terrain->setDrawWire(false);
    auto listener = EventListenerTouchAllAtOnce::create();
    listener->onTouchesMoved = CC_CALLBACK_2(TerrainSimple::onTouchesMoved, this);
    _eventDispatcher->addEventListenerWithSceneGraphPriority(listener, this);
//     add Particle3D for test blend
    auto rootps = PUParticleSystem3D::create("Particle3D/scripts/mp_torch.pu");
    rootps->setCameraMask((unsigned short)CameraFlag::USER1);
    rootps->startParticleSystem();
    
    this->addChild(rootps, 0, 0);
}

std::string TerrainSimple::title() const 
{
    return "Terrain with skirt";
}

std::string TerrainSimple::subtitle() const 
{
    return "Drag to walkThru";
}

void TerrainSimple::onTouchesMoved(const std::vector<cocos2d::Touch*>& touches, cocos2d::Event* event)
{
    float delta = Director::getInstance()->getDeltaTime();
    auto touch = touches[0];
    auto location = touch->getLocation();
    auto PreviousLocation = touch->getPreviousLocation();
    Point newPos = PreviousLocation - location;

    Vec3 cameraDir;
    Vec3 cameraRightDir;
    _camera->getNodeToWorldTransform().getForwardVector(&cameraDir);
    cameraDir.normalize();
    cameraDir.y=0;
    _camera->getNodeToWorldTransform().getRightVector(&cameraRightDir);
    cameraRightDir.normalize();
    cameraRightDir.y=0;
    Vec3 cameraPos=  _camera->getPosition3D();
    cameraPos+=cameraDir*newPos.y*0.5*delta;  
    cameraPos+=cameraRightDir*newPos.x*0.5*delta;
    _camera->setPosition3D(cameraPos);   
}


std::string TerrainWalkThru::title() const 
{
    return "Player walk around in terrain";
}

std::string TerrainWalkThru::subtitle() const 
{
    return "touch to move";
}

TerrainWalkThru::TerrainWalkThru()
{
    auto listener = EventListenerTouchAllAtOnce::create();
    listener->onTouchesBegan = CC_CALLBACK_2(TerrainWalkThru::onTouchesBegan, this);
    listener->onTouchesEnded = CC_CALLBACK_2(TerrainWalkThru::onTouchesEnd, this);
    _eventDispatcher->addEventListenerWithSceneGraphPriority(listener, this);

    Size visibleSize = Director::getInstance()->getVisibleSize();

    //use custom camera
    _camera = Camera::createPerspective(60,visibleSize.width/visibleSize.height,0.1f,200);
    _camera->setCameraFlag(CameraFlag::USER1);
    addChild(_camera);

    Terrain::DetailMap r("TerrainTest/dirt.jpg"),g("TerrainTest/Grass2.jpg",10),b("TerrainTest/road.jpg"),a("TerrainTest/GreenSkin.jpg",20);

    Terrain::TerrainData data("TerrainTest/heightmap16.jpg","TerrainTest/alphamap.png",r,g,b,a,Size(32,32),40.0f,2);
    _terrain = Terrain::create(data,Terrain::CrackFixedType::SKIRT);
    _terrain->setMaxDetailMapAmount(4);
    _terrain->setCameraMask(2);
    _terrain->setDrawWire(false);

    _terrain->setSkirtHeightRatio(3);
    _terrain->setLODDistance(64,128,192);
    _player = Player::create("Sprite3DTest/girl.c3b",_camera,_terrain);
    _player->setCameraMask(2);
    _player->setScale(0.08f);
    _player->setPositionY(_terrain->getHeight(_player->getPositionX(),_player->getPositionZ())+PLAYER_HEIGHT);

    // add Particle3D for test blend
    auto rootps = PUParticleSystem3D::create("Particle3D/scripts/mp_torch.pu");
    rootps->setCameraMask((unsigned short)CameraFlag::USER1);
    rootps->setScale(30.0f);
    rootps->startParticleSystem();
    _player->addChild(rootps);
    
    // add BillBoard for test blend
    auto billboard = BillBoard::create("Images/btn-play-normal.png");
    billboard->setPosition3D(Vec3(0,180,0));
    billboard->setCameraMask((unsigned short)CameraFlag::USER1);
    _player->addChild(billboard);

    auto animation = Animation3D::create("Sprite3DTest/girl.c3b","Take 001");
    if (animation)
    {
        auto animate = Animate3D::create(animation);
        _player->runAction(RepeatForever::create(animate));
    }

    _camera->setPosition3D(_player->getPosition3D()+camera_offset);
    _camera->setRotation3D(Vec3(-45,0,0));

    addChild(_player);
    addChild(_terrain);
}


void TerrainWalkThru::onTouchesBegan(const std::vector<cocos2d::Touch*>& touches, cocos2d::Event* event)
{
}

void TerrainWalkThru::onTouchesEnd(const std::vector<cocos2d::Touch*>& touches, cocos2d::Event* event)
{
    auto touch = touches[0];
    auto location = touch->getLocationInView();
    if(_camera)
    {
        if(_player)
        {
            Vec3 nearP(location.x, location.y, 0.0f), farP(location.x, location.y, 1.0f);

            auto size = Director::getInstance()->getWinSize();
            _camera->unproject(size, &nearP, &nearP);
            _camera->unproject(size, &farP, &farP);
            Vec3 dir = farP - nearP;
            dir.normalize();
            Vec3 collisionPoint(-999,-999,-999);
            bool isInTerrain = _terrain->getIntersectionPoint(Ray(nearP, dir), collisionPoint);
            if (!isInTerrain) 
            {
            _player->idle();
            return;
            }
            dir = collisionPoint - _player->getPosition3D();
            dir.y = 0;
            dir.normalize();
            _player->_headingAngle =  -1*acos(dir.dot(Vec3(0,0,-1)));
            dir.cross(dir,Vec3(0,0,-1),&_player->_headingAxis);
            _player->_targetPos=collisionPoint;
            _player->forward();
        }
    }
}


bool Player::isDone() const
{
    return false;
}


void Player::update(float dt)
{
    auto player = (Sprite3D *)this;
    switch (_playerState)
    {
    case PLAYER_STATE_IDLE:
        break;
    case PLAYER_STATE_FORWARD:
        {
            Vec3 curPos=  player->getPosition3D();
            Vec3 newFaceDir = _targetPos - curPos;
            newFaceDir.y = 0.0f;
            newFaceDir.normalize();
            Vec3 offset = newFaceDir * 25.0f * dt;
            curPos+=offset;
            player->setPosition3D(curPos);
        }
        break;
    case PLAYER_STATE_BACKWARD:
        {
            Vec3 forward_vec;
            player->getNodeToWorldTransform().getForwardVector(&forward_vec);
            forward_vec.normalize();
            player->setPosition3D(player->getPosition3D()-forward_vec*15*dt);
        }
        break;
    case PLAYER_STATE_LEFT:
        {
            player->setRotation3D(player->getRotation3D()+Vec3(0,25*dt,0));
        }
        break;
    case PLAYER_STATE_RIGHT:
        {
            player->setRotation3D(player->getRotation3D()+Vec3(0,-25*dt,0));
        }
        break;
    default:
        break;
    }
    // transform player position to world coord
    auto playerPos = player->getPosition3D();
    auto playerModelMat = player->getParent()->getNodeToWorldTransform();
    playerModelMat.transformPoint(&playerPos);
    Vec3 Normal;
    float player_h = _terrain->getHeight(playerPos.x, playerPos.z,&Normal);
    if (Normal.isZero())//check the player whether is out of the terrain
    {
        player_h = playerPos.y;
    }
    else
    {
        player_h += PLAYER_HEIGHT;
    }
    player->setPositionY(player_h);
    Quaternion q2;
    q2.createFromAxisAngle(Vec3(0,1,0),(float)-M_PI,&q2);
 
    Quaternion headingQ;
    headingQ.createFromAxisAngle(_headingAxis,_headingAngle,&headingQ);
    player->setRotationQuat(headingQ*q2);
    auto vec_offset =Vec4(camera_offset.x,camera_offset.y,camera_offset.z,1);
    vec_offset = player->getNodeToWorldTransform()*vec_offset;
     _cam->setPosition3D(player->getPosition3D() + camera_offset);
    updateState();
}

void Player::turnLeft()
{
    _playerState = PLAYER_STATE_LEFT;
}

void Player::turnRight()
{
    _playerState = PLAYER_STATE_RIGHT;
}

void Player::idle()
{
    _playerState = PLAYER_STATE_IDLE;
}


void Player::forward()
{
    _playerState = PLAYER_STATE_FORWARD;
}

void Player::backward()
{
    _playerState = PLAYER_STATE_BACKWARD;
}

void Player::updateState()
{
    auto player = (Sprite3D * )this;
    switch (_playerState)
    {
    case PLAYER_STATE_FORWARD:
        {
            Vec2 player_pos =Vec2(player->getPositionX(),player->getPositionZ());
            Vec2 targetPos = Vec2(_targetPos.x,_targetPos.z);
            auto dist = player_pos.distance(targetPos);
            if(dist<1)
            {
                _playerState = PLAYER_STATE_IDLE;
            }
        }
        break;
    default:
        break;
    }
}

Player * Player::create(const char * file,Camera * cam,Terrain * terrain)
{
    //
    auto sprite = new (std::nothrow) Player();
    if (sprite && sprite->initWithFile(file))
    {
        sprite->_headingAngle = 0;
        sprite->_playerState = PLAYER_STATE_IDLE;
        sprite->_cam = cam;
        sprite->_terrain = terrain;
        sprite->autorelease();
        sprite->scheduleUpdate();
        return sprite;
    }
    CC_SAFE_DELETE(sprite);
    return nullptr;
}

TerrainWithLightMap::TerrainWithLightMap()
{
    Size visibleSize = Director::getInstance()->getVisibleSize();
    
    //use custom camera
    _camera = Camera::createPerspective(60,visibleSize.width/visibleSize.height,0.1f,800);
    _camera->setCameraFlag(CameraFlag::USER1);
    _camera->setPosition3D(Vec3(-1,1.6f,4));
    addChild(_camera);
    
    Terrain::DetailMap r("TerrainTest/dirt.jpg"),g("TerrainTest/Grass2.jpg"),b("TerrainTest/road.jpg"),a("TerrainTest/GreenSkin.jpg");
    
    Terrain::TerrainData data("TerrainTest/heightmap16.jpg","TerrainTest/alphamap.png",r,g,b,a);
    
    _terrain = Terrain::create(data,Terrain::CrackFixedType::SKIRT);
    _terrain->setLODDistance(3.2f,6.4f,9.6f);
    _terrain->setMaxDetailMapAmount(4);
    addChild(_terrain);
    _terrain->setCameraMask(2);
    _terrain->setDrawWire(false);
    _terrain->setLightMap("TerrainTest/Lightmap.png");
    auto listener = EventListenerTouchAllAtOnce::create();
    listener->onTouchesMoved = CC_CALLBACK_2(TerrainWithLightMap::onTouchesMoved, this);
    _eventDispatcher->addEventListenerWithSceneGraphPriority(listener, this);
    
}
std::string TerrainWithLightMap::title() const
{
    return "Terrain using light map";
}
std::string TerrainWithLightMap::subtitle() const
{
    return "Drag to walkThru";
}
void TerrainWithLightMap::onTouchesMoved(const std::vector<cocos2d::Touch*>& touches, cocos2d::Event* event)
{
    float delta = Director::getInstance()->getDeltaTime();
    auto touch = touches[0];
    auto location = touch->getLocation();
    auto PreviousLocation = touch->getPreviousLocation();
    Point newPos = PreviousLocation - location;
    
    Vec3 cameraDir;
    Vec3 cameraRightDir;
    _camera->getNodeToWorldTransform().getForwardVector(&cameraDir);
    cameraDir.normalize();
    cameraDir.y=0;
    _camera->getNodeToWorldTransform().getRightVector(&cameraRightDir);
    cameraRightDir.normalize();
    cameraRightDir.y=0;
    Vec3 cameraPos=  _camera->getPosition3D();
    cameraPos+=cameraDir*newPos.y*0.5*delta;
    cameraPos+=cameraRightDir*newPos.x*0.5*delta;
    _camera->setPosition3D(cameraPos);
}

File: tests/js-tests/src/TerrainTest/TerrainTest.js
/****************************************************************************
 Copyright (c) 2008-2010 Ricardo Quesada
 Copyright (c) 2011-2012 cocos2d-x.org
 Copyright (c) 2013-2016 Chukong Technologies Inc.
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos2d-x.org

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/
if(cc.sys.isNative)(function(){

var TerrainTestIdx = -1;

var TerrainTestDemo = cc.Layer.extend({
    _title:"",
    _subtitle:"",

    ctor:function () {
        this._super();
    },

    //
    // Menu
    //
    onEnter:function () {
        this._super();

        var label = new cc.LabelTTF(this._title, "Arial", 28);
        this.addChild(label, 100, BASE_TEST_TITLE_TAG);
        label.x = winSize.width / 2;
        label.y = winSize.height - 50;

        var label2 = new cc.LabelTTF(this._subtitle, "Thonburi", 16);
        this.addChild(label2, 101, BASE_TEST_SUBTITLE_TAG);
        label2.x = winSize.width / 2;
        label2.y = winSize.height - 80;

        var item1 = new cc.MenuItemImage(s_pathB1, s_pathB2, this.onBackCallback, this);
        var item2 = new cc.MenuItemImage(s_pathR1, s_pathR2, this.onRestartCallback, this);
        var item3 = new cc.MenuItemImage(s_pathF1, s_pathF2, this.onNextCallback, this);

        item1.tag = BASE_TEST_MENUITEM_PREV_TAG;
        item2.tag = BASE_TEST_MENUITEM_RESET_TAG;
        item3.tag = BASE_TEST_MENUITEM_NEXT_TAG;

        var menu = new cc.Menu(item1, item2, item3);

        menu.x = 0;
        menu.y = 0;
        var width = item2.width, height = item2.height;
        item1.x =  winSize.width/2 - width*2;
        item1.y = height/2 ;
        item2.x =  winSize.width/2;
        item2.y = height/2 ;
        item3.x =  winSize.width/2 + width*2;
        item3.y = height/2 ;

        this.addChild(menu, 102, BASE_TEST_MENU_TAG);
    },

    onRestartCallback:function (sender) {
        var s = new TerrainTestScene();
        s.addChild(restartTerrainTest());
        director.runScene(s);
    },

    onNextCallback:function (sender) {
        var s = new TerrainTestScene();
        s.addChild(nextTerrainTest());
        director.runScene(s);
    },

    onBackCallback:function (sender) {
        var s = new TerrainTestScene();
        s.addChild(previousTerrainTest());
        director.runScene(s);
    },
});

TerrainTestScene = cc.Scene.extend({
    ctor:function () {
        this._super();

        var label = new cc.LabelTTF("Main Menu", "Arial", 20);
        var menuItem = new cc.MenuItemLabel(label, this.onMainMenuCallback, this);

        var menu = new cc.Menu(menuItem);
        menu.x = 0;
        menu.y = 0;
        menuItem.x = winSize.width - 50;
        menuItem.y = 25;
        this.addChild(menu);
    },
    onMainMenuCallback:function () {
        var scene = new cc.Scene();
        var layer = new TestController();
        scene.addChild(layer);
        director.runScene(scene);
    },
    runThisTest:function (num) {
        TerrainTestIdx = (num || num == 0) ? (num - 1) : -1;
        var layer = nextTerrainTest();
        this.addChild(layer);

        director.runScene(this);
    }
});

var TerrainSimple = TerrainTestDemo.extend({
    _title:"Terrain with skirt",
    _subtitle:"Drag to walkthru",
    _camera:null,

    ctor:function(){
        this._super();

        var visibleSize = cc.director.getVisibleSize();

        //use custom camera
        this._camera = new cc.Camera(cc.Camera.Mode.PERSPECTIVE, 60,visibleSize.width/visibleSize.height,0.1,800);
        this._camera.setCameraFlag(cc.CameraFlag.USER1);
        this._camera.setPosition3D(cc.math.vec3(-1, 1.6, 4));
        this.addChild(this._camera);

        var r = jsb.Terrain.detailMap("TerrainTest/dirt.jpg", 35),
            g = jsb.Terrain.detailMap("TerrainTest/Grass2.jpg", 35),
            b = jsb.Terrain.detailMap("TerrainTest/road.jpg", 35),
            a = jsb.Terrain.detailMap("TerrainTest/GreenSkin.jpg", 35);
        var data = jsb.Terrain.terrainData(
            "TerrainTest/heightmap16.jpg",
            "TerrainTest/alphamap.png",
            [r, g, b, a],
            cc.size(32,32),
            2,
            0.1
        );

        var terrain = new jsb.Terrain(data, jsb.Terrain.CrackFixedType.SKIRT);
        terrain.setLODDistance(3.2,6.4,9.6);
        terrain.setMaxDetailMapAmount(4);
        this.addChild(terrain);
        terrain.setCameraMask(2);
        terrain.setDrawWire(false);

        cc.eventManager.addListener({
            event:cc.EventListener.TOUCH_ALL_AT_ONCE,
            onTouchesMoved:this.onTouchesMoved.bind(this)
        }, this);

        var rootps = jsb.PUParticleSystem3D.create("Particle3D/scripts/mp_torch.pu");
        rootps.setCameraMask(2);
        rootps.startParticleSystem();

        this.addChild(rootps, 0, 0);
    },

    onTouchesMoved:function(touches, event){
        var delta = cc.director.getDeltaTime();
        var touch = touches[0];
        var location = touch.getLocation();
        var previousLocation = touch.getPreviousLocation();
        var newPos = cc.p(previousLocation.x - location.x, previousLocation.y - location.y);

        var m = this._camera.getNodeToWorldTransform3D();
        var cameraDir = cc.math.vec3(-m[8], -m[9], -m[10]);
        cameraDir.normalize();
        cameraDir.y = 0;

        var cameraRightDir = cc.math.vec3(m[0], m[1], m[2]);
        cameraRightDir.normalize();
        cameraRightDir.y = 0;

        var cameraPos = this._camera.getPosition3D();
        cameraPos.x += cameraDir.x * newPos.y * 0.5 *delta + cameraRightDir.x * newPos.x * 0.5 *delta;
        cameraPos.y += cameraDir.y * newPos.y * 0.5 *delta + cameraRightDir.y * newPos.x * 0.5 *delta;
        cameraPos.z += cameraDir.z * newPos.y * 0.5 *delta + cameraRightDir.z * newPos.x * 0.5 *delta;
        this._camera.setPosition3D(cameraPos);
    }
});

var TerrainWalkThru = (function(){
    var PlayerState = {
        LEFT : 0,
        RIGHT : 1,
        IDLE : 2,
        FORWARD : 3,
        BACKWARD : 4
    };

    var PLAYER_HEIGHT = 0;
    var camera_offset = cc.math.vec3(0, 45, 60);

    var Player = jsb.Sprite3D.extend({
        _targetPos:null,
        _headingAxis:null,
        _headingAngle:0,
        _playerState:PlayerState.IDLE,
        _camera:null,
        _terrain:null,

        ctor:function(file, camera, terrain){
            this._super(file);
            this._camera = camera;
            this._terrain = terrain;
            this.scheduleUpdate();
        },

        update:function(dt){
            switch(this._playerState){
            case PlayerState.IDLE:
                break;
            case PlayerState.FORWARD:
                var curPos = this.getPosition3D();
                var newFaceDir = cc.math.vec3Sub(this._targetPos, curPos);
                newFaceDir.y = 0;
                newFaceDir.normalize();
                curPos.x += newFaceDir.x * 25 * dt;
                curPos.y += newFaceDir.y * 25 * dt;
                curPos.z += newFaceDir.z * 25 * dt;
                this.setPosition3D(curPos);
                break;
            case PlayerState.BACKWARD:
                break;
            case PlayerState.LEFT:
                this.setRotation3D(cc.math.vec3Add(this.getRotation3D(), cc.math.vec3(0, 25*dt, 0)));
                break;
            case PlayerState.RIGHT:
                this.setRotation3D(cc.math.vec3Add(this.getRotation3D(), cc.math.vec3(0, -25*dt, 0)));
                break;
            default:
                break;
            }

            var player_h = this._terrain.getHeight(this.getPositionX(), this.getVertexZ());
            this.y = player_h + PLAYER_HEIGHT;
            this._camera.setPosition3D(cc.math.vec3Add(this.getPosition3D(), camera_offset));

            if(this._headingAxis){
                var quat = cc.math.quaternion(cc.math.vec3(0, 1, 0), -Math.PI);
                var headingQ = cc.math.quaternion(this._headingAxis, this._headingAngle);
                this.setRotationQuat(cc.math.quatMultiply(quat, headingQ));
            }

            this.updateState();
        },

        turnLeft:function(){
            this._playerState = PlayerState.LEFT;
        },

        turnRight:function(){
            this._playerState = PlayerState.RIGHT;
        },

        forward:function(){
            this._playerState = PlayerState.FORWARD;
        },

        backward:function(){
            this._playerState = PlayerState.BACKWARD;
        },

        idle:function(){
            this._playerState = PlayerState.IDLE;
        },

        updateState:function(){
            if(this._playerState === PlayerState.FORWARD){
                var playerPos = cc.p(this.getPositionX(), this.getVertexZ());
                var targetPos = cc.p(this._targetPos.x, this._targetPos.z);
                var dist = cc.pDistance(playerPos, targetPos);
                if(dist < 1){
                    this._playerState = PlayerState.IDLE;
                }
            }
        }
    });

    return TerrainTestDemo.extend({
        _title:"Player walk around in terrain",
        _subtitle:"touch to move",
        _camera:null,
        _player:null,
        _terrain:null,

        ctor:function(){
            this._super();

            cc.eventManager.addListener({
                event:cc.EventListener.TOUCH_ALL_AT_ONCE,
                onTouchesEnded:this.onTouchesEnded.bind(this)
            }, this);

            var visibleSize = cc.director.getVisibleSize();
            this._camera = new cc.Camera(cc.Camera.Mode.PERSPECTIVE, 60, visibleSize.width/visibleSize.height, 0.1, 200);
            this._camera.setCameraFlag(cc.CameraFlag.USER1);
            this.addChild(this._camera);

            var r = jsb.Terrain.detailMap("TerrainTest/dirt.jpg"),
                g = jsb.Terrain.detailMap("TerrainTest/Grass2.jpg", 10),
                b = jsb.Terrain.detailMap("TerrainTest/road.jpg"),
                a = jsb.Terrain.detailMap("TerrainTest/GreenSkin.jpg", 20);
            var data = jsb.Terrain.terrainData(
                "TerrainTest/heightmap16.jpg",
                "TerrainTest/alphamap.png",
                [r, g, b, a],
                cc.size(32,32),
                40.0,
                2
            );

            this._terrain = new jsb.Terrain(data, jsb.Terrain.CrackFixedType.SKIRT);
            this._terrain.setLODDistance(64, 128, 192);
            this._terrain.setMaxDetailMapAmount(4);
            this._terrain.setCameraMask(2);
            this._terrain.setDrawWire(false);
            this._terrain.setSkirtHeightRatio(3);
            this.addChild(this._terrain);

            this._player = new Player("Sprite3DTest/girl.c3b", this._camera, this._terrain);
            this._player.setCameraMask(2);
            this._player.setScale(0.08);
            this._player.y = this._terrain.getHeight(this._player.x, this._player.getVertexZ()) + PLAYER_HEIGHT;

            var animation = new jsb.Animation3D("Sprite3DTest/girl.c3b", "Take 001");
            if(animation){
                var animate = new jsb.Animate3D(animation);
                this._player.runAction(cc.repeatForever(animate));
            }

            this.addChild(this._player);

            this._camera.setPosition3D(cc.math.vec3Add(this._player.getPosition3D(), camera_offset));
            this._camera.setRotation3D(cc.math.vec3(-45, 0, 0));
        },

        onTouchesEnded:function(touches, event){
            var touch = touches[0];
            var location = touch.getLocationInView();
            if(this._camera && this._player){
                var nearP = cc.math.vec3(location.x, location.y, 0);
                var farP = cc.math.vec3(location.x, location.y, 1);

                nearP = this._camera.unproject(nearP);
                farP = this._camera.unproject(farP);

                var dir = cc.math.vec3Sub(farP, nearP);
                dir.normalize();
                var rayStep = cc.math.vec3(15*dir.x, 15*dir.y, 15*dir.z);
                var rayPos = nearP;
                var rayStartPosition = nearP;
                var lastRayPosition = rayPos;
                rayPos.x += rayStep.x;
                rayPos.y += rayStep.y;
                rayPos.z += rayStep.z;

                // Linear search - Loop until find a point inside and outside the terrain Vector3
                var height = this._terrain.getHeight(rayPos.x, rayPos.z);
                while(rayPos.y > height){
                    lastRayPosition = rayPos;
                    rayPos.x += rayStep.x;
                    rayPos.y += rayStep.y;
                    rayPos.z += rayStep.z;
                    height = this._terrain.getHeight(rayPos.x, rayPos.z);
                }

                var startPosition = lastRayPosition;
                var endPosition = rayPos;

                for(var i = 0; i < 32; ++i){
                    // Binary search pass
                    var middlePoint = cc.math.vec3((startPosition.x+endPosition.x)*0.5, (startPosition.y+endPosition.y)*0.5, (startPosition.z+endPosition.z)*0.5);
                    if(middlePoint.y < height)
                        endPosition = middlePoint;
                    else
                        startPosition = middlePoint;
                }

                var collisionPoint = cc.math.vec3((startPosition.x+endPosition.x)*0.5, (startPosition.y+endPosition.y)*0.5, (startPosition.z+endPosition.z)*0.5);
                dir = cc.math.vec3Sub(collisionPoint, this._player.getPosition3D());
                dir.y = 0;
                dir.normalize();
                this._player._headingAngle = -1 * Math.acos(cc.math.vec3Dot(dir, cc.math.vec3(0, 0, -1)));
                this._player._headingAxis = cc.math.vec3Cross(dir, cc.math.vec3(0, 0, -1));
                this._player._targetPos = collisionPoint;
                this._player.forward();
            }
        }
    });
})();
//
// Flow control
//
var arrayOfTerrainTest = [
    TerrainSimple,
    TerrainWalkThru
];

var nextTerrainTest = function () {
    TerrainTestIdx++;
    TerrainTestIdx = TerrainTestIdx % arrayOfTerrainTest.length;

    return new arrayOfTerrainTest[TerrainTestIdx ]();
};
var previousTerrainTest = function () {
    TerrainTestIdx--;
    if (TerrainTestIdx < 0)
        TerrainTestIdx += arrayOfTerrainTest.length;

    return new arrayOfTerrainTest[TerrainTestIdx ]();
};
var restartTerrainTest = function () {
    return new arrayOfTerrainTest[TerrainTestIdx ]();
};

})();
